### ISSUE TITLE
Use editor.formatOnPaste setting

### ISSUE THREAD

**Issue by rightaway** [VSCodeVim/Vim#1269]
The new `editor.formatOnPaste` setting in vscode 1.9 works when pasting with `ctrl-v` with not in vim mode's `p`, which should use that setting too to automatically format when pasting.

**Comment by rebornix** [comment]
The `formatOnPaste` feature listens to browser's `paste` event however our `p` is not a real paste. That's why Code's native `formatOnPaste` doesn't kick in.

A proper solution is 

* Check `editor.formatOnPaste` when running `p`
* If it's true, run format selection on `p`'s pasted content.
* Set correct final selection.

It's like re-implementing `formatOnPaste` again in Vim. Since I did that in Code, I'd like to leave this to anyone who wants to contribute :)

**Comment by rebornix** [comment]
@johnfn yeah of course. The only thing I'm not sure about is whether users (majority) wants to have this feature in Vim.

**Comment by johnfn** [comment]
That's a very good point actually. It's good to focus on issues with the most thumbs up.

**Comment by ronakg** [comment]
Is there a workaround for this?

**Comment by J-Fields** [comment]
>A proper solution is
Check `editor.formatOnPaste` when running `p`
If it's true, run format selection on `p`'s pasted content.
Set correct final selection.

I'll mention that I prototyped this at one point and it wasn't a great experience. What we *really* want is VSCode to expose its `formatOnPaste` implementation, since then from our side it would be atomic. Still, a sort of ugly solution is better than none.

**Comment by jim-alexander** [comment]
This seemed to work at a glance: 

```JSON
{ "before": ["p"], "commands": ["editor.action.formatDocument"], "after": ["p"] },
```

**Comment by diminutivesloop** [comment]
@jim-alexander One downside of that solution is that it will format the entire document which may not always be desirable as it could result in larger diffs than expected. `editor.formatOnPaste` formats only the pasted content.
